### PR TITLE
Generate node visitor mappings automatically

### PR DESCRIPTION
**Pull request by jdevera**
# 📖 What this PR does

Fixes: #35

Before this PR, in order to automatically generate a visitor for the Mypy AST, the mapping between visitor method name and the type of the node being visited is stored in a dictionary with hardcoded values.

This PR introduces the automatic generation of the same mapping.

The docstring of the `visitor.method_mapping` module contains a full explanation of why this change was not trivial, reproduced here:

> This module provides a mapping between a method name in a Visitor or Mypy's ASTs and the type of the Node it is meant to visit.
>
> The purpose of this is to enable the dynamic creation of visitors to the AST. This information is surprisingly hard to obtain programmatically. The approach here is to explore all the methods of an existing `Visitor` class in Mypy: `mypy.traverser.TraverserVisitor` and obtain the type annotation for their first (non-self) parameter.
>
> This is further complicated by the fact that Mypy loads by default as compiled code, and typing information for methods if thus not available. Here we use a trick found [here on Stack Overflow](https://stackoverflow.com/a/68685189/) to create a context manager that temporarily forces a preference for pure python modules when importing.
>
> So roughly, we do this:
>
> 1. Import the mypy things we need
> 2. Capture the globals (so that we can resolve the strigified type annotations to the correct types)
> 3. Clear the mypy imported modules
> 4. Import them again with their pure python versions
> 5. Inspect the Visitor to get the type names, but resolve them using the captured globals (from the native versions)
> 6. Restore the native mypy implementations

# 🤔 Some thoughts on the approach

After all of this, which has yielded enormous amounts of fun, I am not 100% convinced I even like this approach. I have an intuition that it is a bit brittle (the whole unloading modules to reload them so one can inspect types at runtime is a bit suspicious).

I have been thinking perhaps harcoding is not so bad after all, but having a dictionary of method names to node types does not provide many guarantees that things will continue to work when mypy changes, or that we will be notified.

I started to think about going all-in in the opposite direction: Create a visitor blueprint, explicitly overriding all visitor methods, and then write tests and use linters to let us know if something changes upstream.

It has the added benefit of being straightforward in the code, although longer, for sure.

## Alternative 1 (no code generation)

It could look like this:

```python
class RefurbVisitor(TraverserVisitor):

    def __init__(self, checks: dict[mypy.nodes.Node, list[Check]]):
        super().__init__()
        self.checks = checks
        self.errors = []

    def run_checks(self, node):
        for check in self.checks.get(type(node), []):
            check(node, self.errors)

    def visit_mypy_file(self, o: mypy.nodes.MypyFile):
        super().visit_mypy_file(o)
        self.run_checks(o)

    def visit_block(self, o: mypy.nodes.Block):
        super().visit_block(o)
        self.run_checks(o)
```

This does not need to generate any methods or code for them beyond loading the checkers in a dictionary keyed on the node type, like today.

And there can be tests in place to ensure all `visit_` methods are overridden. So we'd get the warning in case of a change of mypy version.

## Alternative 2 (some code generation)

Or it could like like this:

```python
class VisitorBlueprint(TraverserVisitor):

    def visit_mypy_file(self, o: mypy.nodes.MypyFile):
        pass

    def visit_block(self, o: mypy.nodes.Block):
        pass

    def visit_func(self, o: mypy.nodes.FuncItem):
        pass
```

In this case, the blueprint can be used as the MAPPING is used today, only this one is a living version that declares its parent and linters can better check its correctness.

Then we can iterate on the methods to generate the code for them when building the ultimate `RefurbVisitor`.

# 🥇 My Recommendation

Even when this whole thing started with a TODO comment that said:
```
# TODO: Dynamically generate this somehow instead of hardcoding it
```

I think there is more virtue in being explicit, and I would recommend alternative 1 (and I'm happy to contribute that).

After this PR is here, I feel a bit like this about it:

 But it is ultimately up to you @dosisod :)

### CODE REVIEW

**Comment by dosisod** [comment]
Thank you for submitting such a well-documented and thought out pull request for this! I will make sure to take a look at this tomorrow. I agree that there are pros and cons to each of the strategies you mention, so we will have to weight each one and see which one strikes the best balance between speed, maintainablity, and robustness.

File: refurb/visitor/method_mapping.py [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+from contextlib import contextmanager
+from dataclasses import dataclass
+from importlib.abc import Finder
+from importlib.machinery import FileFinder
+from types import FunctionType
+from typing import Callable, Any
+import mypy
+import mypy.traverser
+
+VISITOR_NAME_TO_NODE_TYPE_MAPPING = dict()
```
[the remark this answers is not in the document]
Comment (dosisod) [reply refurb/visitor/method_mapping.py:43]: Also, I think `VISITOR_NAME_TO_NODE_TYPE_MAPPINGS` is too verbose, perhaps `VISITOR_NODE_MAPPINGS` or something similar?

**Review by dosisod — CHANGES_REQUESTED** [review]
Sorry for the delay on this review! I agree that this does feel a bit "suspicious", particularly with the loading and unloading of `sys.modules` at runtime. I think I might have found a more isolated approach using `importlib`. It is not fully working, but I thought I would share it now instead of reviewing a bunch of code only for it to change drastically.

In short:

* We load the `mypy.traverser` module
* Find the filepath it is loaded from, and then load the `traverser.py` file that is in the same directory as it
* Load and inspect `TraverserVisitor` as usual

The code:

```python
import importlib.util
import typing
from pathlib import Path

spec = importlib.util.find_spec("mypy.traverser")

if spec is not None:
    assert spec.origin

    path = str(Path(spec.origin).with_name("traverser.py"))

    spec = importlib.util.spec_from_file_location("mypy.traverser", path)
    assert spec

    module = importlib.util.module_from_spec(spec)
    spec.loader.exec_module(module)

    import inspect

    members = inspect.getmembers(
        module.TraverserVisitor,
        lambda o: inspect.isfunction(o) and o.__name__.startswith("visit_"),
    )

    for method_name, method in members:
        method_params = list(inspect.signature(method).parameters.values())
        param_name = method_params[1].name

        # method_types = typing.get_type_hints(method, globalns=globals())

        print(method_name, method_params)
```

Do note the commented out `get_type_hints` line, which fails when uncommented. I think this is due to the `mypy`/`mypy.patterns`/`mypy.nodes` modules not being loaded. I assume this will be an easy fix, though I may be mistaken.

Great job on the work thus far!

**Comment by jdevera** [comment]
This line is trickier than it seems:
```
# method_types = typing.get_type_hints(method, globalns=globals())
```
Mainly because, in order to resolve the node types, it uses the contents of `globalns` and, at the time of the call, these would include only what has been imported.

( 15 mins later :P )

However, because of the way you went about importing, `mypy.traverser` is nowhere in `sys.modules` at the time, so I tried just importing `mypy.traverser` and then passing `vars(mypy.traverser) as the `globalns` parameter and ... IT WORKS! 🎉 

```python
import mypy.traverser

...

method_types = typing.get_type_hints(method, globalns=vars(mypy.traverser))
```

And then later:

```
mypy.nodes.YieldFromExpr is method_types['o']
Out[26]: True
```

I'll give it another round. ♻️ 

P.S.: Also, infinite thank-yous for being so responsive, you cannot be seriously apologising for delays in getting back to me on this, haha, I've had pull requests open for months without a word, this is a treat! Especially considering the fact that this is a lot of geeking out about python and not so crucial for "the product". I appreciate it.

**Comment by jdevera** [comment]
News, and they aren't good, I'm afraid. Some of the methods of a visitor are not overridden by the `TraverseVisitor` so when we load only that module (`mypy.traverser`) its internal imports are not importing the python versions. So when the class is inspected, the inherited methods from `NodeVisitor[T]` are coming back as our good old friends `methoddescriptor`, meaning no type info to extract.

**Comment by jdevera** [comment]
I'm now thinking more in terms of not changing the current behaviour 😓 , because I have been playing with alternatives and none are great:

1. A Visitor that uses `__getattribute__` to generate the methods as they are called.
2. A Visitor that explicitly declares all methods but has no content in them, and then uses a class decorator add all the logic to each method
3. A Visitor that explicitly declares all methods and their contents, no generation, plain python code, but very repetitive

The first one has additional issues with some of the visitor methods passing down the same node to the next and relying on inheritance to make things work. It is also impossible to get a list of node types for the generator command with this one.

With the other two, it's easy to extract the node types with some inspection, but they are very verbose.

I want to go back to proposing this:

Shred this PR 😆 . I send another one that uses the import trickery from this PR **but in tests**. The test verifies that the mapping contains precisely what it needs to (today, it has an extra entry), and we avoid all the shenanigans when running the tool while ensuring a very test failure will let us know if a change in mypy has brought in deviations.

**Pull request closed without merging**